app/lib/account/backend.dart

```diff
@@ -191,55 +220,93 @@ Future<AuthenticatedAgent> requireAuthenticatedAgent(
   if (token == null || token.isEmpty) {
     throw AuthenticationException.authenticationRequired();
   }
-  final authenticatedService = await _tryAuthenticateGithubAction(token);
-  if (authenticatedService != null) {
-    return authenticatedService;
+  final authenticatedServiceAgent =
+      await _tryAuthenticateServiceAgent(token, source: source);
+
+  if (authenticatedServiceAgent != null) {
+    return authenticatedServiceAgent;
   } else {
     return AuthenticatedUser(await requireAuthenticatedUser(source: source));
   }
 }
 
-Future<AuthenticatedGithubAction?> _tryAuthenticateGithubAction(
-    String token) async {
+Future<AuthenticatedAgent?> _tryAuthenticateServiceAgent(
+  String token, {
+  AuthSource? source,
+}) async {
   if (!JsonWebToken.looksLikeJWT(token)) {
     return null;
   }
   final idToken = JsonWebToken.tryParse(token);
   if (idToken == null) {
     return null;
   }
-  if (idToken.payload.iss != GitHubJwtPayload.githubIssuerUrl) {
-    return null;
+
+  if (idToken.payload.iss == GitHubJwtPayload.issuerUrl) {
+    // At this point we have confirmed that the token is a JWT token
+    // issued by GitHub. If there is an issue with the token, the
+    // authentication should fail without any fallback.
+    final payload = await _verifyAndParseToken(
+      idToken,
+      openIdDataFetch: fetchGithubOpenIdData,
+      payloadTryParse: GitHubJwtPayload.tryParse,
+    );
+
+    return AuthenticatedGithubAction(
+      idToken: idToken,
+      payload: payload,
+    );
   }
 
-  // At this point we have confirmed that the token is a JWT token
-  // issued by GitHub. If there is an issue with the token, the
-  // authentication should fail without any fallback.
-  return await _authenticateGithubAction(idToken);
+  if (idToken.payload.iss == GoogleCloudServiceAccountJwtPayload.issuerUrl &&
+      source == AuthSource.client &&
+      idToken.payload.aud.length == 1 &&
+      idToken.payload.aud.single ==
+          activeConfiguration.automatedPublishingAudience) {
+    // As the uploader token's audience and the admin token's issuer and also
+    // their audience is the same, we only parse it as a non-user token, when
+    // the authentication source is from the pub client app (e.g. uploading a
+    // new package).
+    // At this point we don't fall back to authenticating the token as a user.
+    final payload = await _verifyAndParseToken(
+      idToken,
+      openIdDataFetch: fetchGoogleCloudOpenIdData,
+      payloadTryParse: GoogleCloudServiceAccountJwtPayload.tryParse,
+    );
```

Comment:
We could also use the tokeninfo end-point, then we don't have to do the crypto here..